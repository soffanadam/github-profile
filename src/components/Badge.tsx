import React from 'react'

export type BadgeProps = {
  dataTestID?: string
}

export const Badge: React.FC<BadgeProps> = ({ dataTestID, children }) => {
  return (
    <span
      data-testid={dataTestID}
      className="inline-flex items-center py-0.5 px-2.5 text-gray-500 bg-white rounded border"
    >
      {children}
    </span>
  )
}
