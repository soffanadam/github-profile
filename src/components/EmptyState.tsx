import { classNames } from '@/utils'
import React from 'react'

export type EmptyStateProps = {
  dataTestID?: string
  className?: string
}

export const EmptyState: React.FC<EmptyStateProps> = ({
  children,
  dataTestID,
  className
}) => {
  return (
    <div
      data-testid={dataTestID}
      className={classNames('py-3 px-5 text-xl', className)}
    >
      {children}
    </div>
  )
}
