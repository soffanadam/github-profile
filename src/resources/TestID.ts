export enum TestID {
  SEARCH_FORM = 'search-form',
  SEARCH_FORM_INPUT = 'search-form-input'
}
