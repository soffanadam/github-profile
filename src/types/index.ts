// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface PlainObject<T = any> {
  [key: string]: T
}

//==============================================================================
// Model
//==============================================================================

export interface User {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [anyProp: string]: any
}

export interface Repo {
  id: number
  name: string
  description?: string
  stargazers_count: number
  forks_count: number
  owner: User
}

export interface Content {
  name: string
  path: string
  type: string
  html_url?: string
}

export interface ReadMe {
  name: string
  content: string
  encoding: string
}

//==============================================================================
// State
//==============================================================================

export interface UserState {
  user: User | null
  error: string
  loading: boolean
  searching: boolean
}

export interface ReposState {
  repos: Repo[]
  page: number
  hasMore: boolean
  error: string
  loading: boolean
}

export interface RepoState {
  repo: Repo | null
  error: string
  loading: boolean
}

export interface ContentsState {
  path: string
  contents: Content[]
  readMe: ReadMe | null
  error: string
  loading: boolean
}

export interface RootState {
  userState: UserState
  reposState: ReposState
  repoState: RepoState
  contentsState: ContentsState
}

//==============================================================================
// Actions Payload
//==============================================================================

export interface GetRepoPayload {
  userName: string
  repoName: string
}

//==============================================================================
// API
//==============================================================================

export class ApiError extends Error {
  response: ApiErrorResponse | null

  constructor(message: string, response: ApiErrorResponse | null = null) {
    super(message)
    this.response = response
  }
}

export interface ApiErrorResponse {
  status?: number
  body?: unknown
}
