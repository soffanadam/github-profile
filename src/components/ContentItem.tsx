import { Content, ContentType } from '@/types'
import React from 'react'
import { FcFile, FcFolder, FcLink, FcPackage } from 'react-icons/fc'

export const ICON_TEST_ID = 'icon'

export type ContentItemProps = {
  dataTestID?: string
  content: Content
}

export const ContentItem: React.FC<ContentItemProps> = ({
  dataTestID,
  content
}) => {
  return (
    <span
      data-testid={dataTestID}
      className="flex items-center py-2 px-5 hover:bg-gray-50 border-b"
    >
      <span className="w-8">
        {content.type == ContentType.DIR ? (
          <FcFolder data-icon={ContentType.DIR} />
        ) : content.type == ContentType.SYMLINK ? (
          <FcLink data-icon={ContentType.SYMLINK} />
        ) : content.type == ContentType.SUBMODULE ? (
          <FcPackage data-icon={ContentType.SUBMODULE} />
        ) : (
          <FcFile data-icon={ContentType.FILE} />
        )}
      </span>
      {content.name}
    </span>
  )
}
