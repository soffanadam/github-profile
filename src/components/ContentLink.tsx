import { Content, ContentType, PlainObject } from '@/types'
import React from 'react'
import { Link } from 'react-router-dom'

export type ContentLinkProps = {
  dataTestID?: string
  repoUrl: string
  content: Content
}

export const ContentLink: React.FC<ContentLinkProps & PlainObject> = ({
  dataTestID,
  repoUrl,
  content,
  children,
  ...rest
}) => {
  if (content.type === ContentType.DIR) {
    return (
      <Link
        data-testid={dataTestID}
        to={`${repoUrl}/${content.path}`}
        {...rest}
      >
        {children}
      </Link>
    )
  }

  if (content.html_url) {
    return (
      <a
        data-testid={dataTestID}
        href={content.html_url}
        target="_blank"
        rel="noreferrer"
        {...rest}
      >
        {children}
      </a>
    )
  }

  return (
    <div data-testid={dataTestID} {...rest}>
      {children}
    </div>
  )
}
