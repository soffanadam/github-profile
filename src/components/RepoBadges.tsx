import { Repo } from '@/types'
import React from 'react'
import { FiStar, FiGitBranch } from 'react-icons/fi'
import { Badge } from './Badge'

export type RepoBadgesProps = {
  repo: Repo
}

export const RepoBadges: React.FC<RepoBadgesProps> = ({ repo }) => {
  return (
    <div className="flex mt-2 space-x-3">
      <Badge>
        <FiStar className="mr-2" />
        <span data-stargazers-count>{repo.stargazers_count}</span>
      </Badge>
      <Badge>
        <FiGitBranch className="mr-2" />
        <span data-forks-count>{repo.forks_count}</span>
      </Badge>
    </div>
  )
}
