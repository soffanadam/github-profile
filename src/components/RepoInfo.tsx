import { Repo } from '@/types'
import React from 'react'
import { Link } from 'react-router-dom'

export type RepoInfoProps = {
  dataTestID?: string
  repo: Repo
}

export const RepoInfo: React.FC<RepoInfoProps> = ({ dataTestID, repo }) => {
  return (
    <Link
      data-testid={dataTestID}
      to={`/${repo.owner.login}/${repo.name}`}
      className="group block"
    >
      <span
        data-testid="repo-info-name"
        className="inline-flex pb-1 mb-2 text-2xl text-indigo-500 border-b border-transparent group-hover:border-indigo-300"
      >
        {repo.name}
      </span>
      {repo.description && (
        <p data-testid="repo-info-description" className="mb-5">
          {repo.description}
        </p>
      )}
    </Link>
  )
}
