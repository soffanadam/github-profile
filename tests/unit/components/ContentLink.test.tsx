import React from 'react'
import { render } from '@testing-library/react'
import { ContentLink, ContentLinkProps } from '@/components/ContentLink'
import { TestID } from '@/resources/TestID'
import { BrowserRouter } from 'react-router-dom'
import { withContentLinkProps } from './utils'
import { ContentType } from '@/types'

const TEXT = 'text'

const renderComponent = (props: ContentLinkProps) => {
  return render(
    <BrowserRouter>
      <ContentLink {...props}>{TEXT}</ContentLink>
    </BrowserRouter>
  )
}

describe('<ContentLink />', () => {
  it('renders correctly', () => {
    const props = withContentLinkProps(ContentType.DIR)
    const component = renderComponent(props)
    expect(component).toBeTruthy()

    const testElement = component.getByTestId(TestID.CONTENT_LINK)
    expect(testElement).toHaveTextContent(TEXT)
    expect(testElement).toMatchSnapshot()
  })

  it('linked to path if a directory', () => {
    const props = withContentLinkProps(ContentType.DIR)
    const { getByTestId } = renderComponent(props)
    expect(getByTestId(TestID.CONTENT_LINK)).toHaveAttribute(
      'href',
      `${props.repoUrl}/${props.content.path}`
    )
  })

  it('linked to external url if non directory and has html_url', () => {
    const props = withContentLinkProps(ContentType.FILE)
    const { getByTestId } = renderComponent(props)
    expect(getByTestId(TestID.CONTENT_LINK)).toHaveAttribute(
      'href',
      props.content.html_url
    )
  })

  it('has no link', () => {
    const props = withContentLinkProps(ContentType.FILE)
    delete props.content.html_url

    const { getByTestId } = renderComponent(props)
    expect(getByTestId(TestID.CONTENT_LINK)).not.toHaveAttribute('href')
  })
})
