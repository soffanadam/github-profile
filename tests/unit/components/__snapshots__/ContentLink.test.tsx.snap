// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentLink /> renders correctly 1`] = `
<a
  data-testid="content-link"
  href="/repo/pathtocontent"
>
  text
</a>
`;
